src/_P003_Pulse.ino

```diff
@@ -292,60 +402,249 @@ boolean Plugin_003(byte function, struct EventStruct *event, String& string)
+            bitClear(P003_processingFlags, event->TaskIndex);
+          }
+          else
+          // we found unexpected different pin state
+          {
+            #ifdef P003_PULSE_STATISTIC
+            P003_Step3NOKcounter[event->TaskIndex]++;
+            #endif // P003_PULSE_STATISTIC
+            // ignore spike from previous step. It is regarded as spike within previous=current signal. Again try to detect stable signal
+            P003_lastCheckState [event->TaskIndex] = pinState;    // now trust the previous=new state
+            Scheduler.setPluginTaskTimer(PCONFIG(P003_IDX_DEBOUNCETIME) >> 1, event->TaskIndex, P003_PROCESSING_STEP_2);
+          }
+          break;  
+        }
+        default:
+        {
+          log = F("_P003:PLUGIN_TIMER_IN: Invalid processingStep: ");
+          log += processingStep;
+          addLog(LOG_LEVEL_ERROR, log);
+          break;
+        }
+      }
+      break;
+    }
+
   }
   return success;
 }
 
+#ifdef P003_PULSE_STATISTIC
+/*********************************************************************************************\
+* write statistic counters to logfile
+\*********************************************************************************************/
+void doStatisticLogging(taskIndex_t TaskIndex, String logPrefix)
+{
+  // Statistic to logfile. E.g: ... [123|111|100/5|80/3/4|40] [12243|3244]
+  String log = logPrefix;
+  log += F("Stats[step0|1|2|3|tot(ok/nok/ign)] [lo|hi]: [");
+  log += P003_Step0counter[TaskIndex];        log += "|";
+  log += P003_Step1counter[TaskIndex];        log += "|";
+  log += P003_Step2OKcounter[TaskIndex];      log += "/";
+  log += P003_Step2NOKcounter[TaskIndex];     log += "|";
+  log += P003_Step3OKcounter[TaskIndex];      log += "/";
+  log += P003_Step3NOKcounter[TaskIndex];     log += "/";
+  log += P003_Step3IGNcounter[TaskIndex];     log += "|";
+  log += P003_pulseTotalCounter[TaskIndex];   log += "] [";
+  log += P003_pulseLowTime[TaskIndex]/1000L;  log += "|";
+  log += P003_pulseHighTime[TaskIndex]/1000L; log += "]";
+  addLog(P003_PULSE_STATS_LOG_LEVEL, log);
+}
+#endif // P003_PULSE_STATISTIC
+
 /*********************************************************************************************\
 * Check Pulse Counters (called from irq handler)
 \*********************************************************************************************/
-void Plugin_003_pulsecheck(byte Index)
+void Plugin_003_pulsecheck(byte taskID)
 {
   noInterrupts(); // s0170071: avoid nested interrups due to bouncing.
-
-  const uint64_t PulseTime = getMicros64() - Plugin_003_pulseTimePrevious[Index];
-  int pinState;
-
-  // Mx: 2021-01: added processing for new mode types PULSE_LOW and PULSE_HIGH
-  // if Mode Type is P003_MODE_TYPE_PULSE_LOW or P003_MODE_TYPE_PULSE_HIGH
-  if ( (Settings.TaskDevicePluginConfig[Index][2] & P003_MODE_TYPE_MODE_MASK) != 0 )
+   
+  if ( (Settings.TaskDevicePluginConfig[taskID][P003_IDX_MODETYPE] & PULSE_MODE_MASK) == 0 )
```

Comment:
I am not sure about this part.
The `Settings` struct is not volatile, so this could cause issues.
Isn't it better to store this as a single `volatile uint32_t` and bitset per task on the `PLUGIN_INIT` call?

Reply:
My understanding is, and I double checked it in the reference, that "volatile" is needed in **regular code** for variables that are **changed** in interrupt handlers, so that the value is always read from RAM and not from registers. Here we are a) inside the ISR, b) are in the noInterupts() part c) variable is not changed in ISR. So the variable cannot be changed, while this code section is executed.
What do you think?

Reply:
Well I am not entirely sure.
The simple fact that it is running stable for a while does suggest you're right about the need for volatile declared values only if you need to change them (which makes sense)
Also the `Settings` object is for sure present in RAM and not on flash or RTC, so this also does not pose any extra uncertainties.
I have not yet found any clear documentation on what exactly are (all) the "rules" for using code and variables in ISR code.
It is all scattered around in lots of places.

Reply:
But in this context, I am wondering, if the writing in the legacy code section inside the ISR of uint64_t in
`P003_currentStableStartTime[taskID]  = getMicros64();`
is a good idea, because when reading the double-word variable in the "normal" (non-ISR) code could be interrupted by the ISR and half the word could be from before and after the interrupt.
I think that generally around the processing of a uint64_t variable, which is changed in an ISR, there should be a (no)Interupts(). However it is actually not needed here, because for the legacy edge mode types, the variable is solely read and written inside the ISR inside the (no)Interupts() section and in the new PULSE mode types it's only used in the "normal" code.
So this should be save.

Reply:
> I have not yet found any clear documentation on what exactly are (all) the "rules" for using code and variables in ISR code.
> It is all scattered around in lots of places.

You are right. Some rules appear here and there. I found the following is a very good summary from an experienced programmer and it not only lists the rules, it also explains the the background and gives also good solutions.
[http://gammon.com.au/interrupts](url)